### ISSUE TITLE
Guard jQuery.extend against non-object arguments (especially strings)

### ISSUE THREAD

**Issue by felixhayashi** [jquery/jquery#2629]
Hi,

When calling the extend method and passing a string as argument, the string is split and character positions are added as property keys with the corresponding characters as values.

![2015-10-04 12 56 10](https://cloud.githubusercontent.com/assets/4307137/10267535/9af4829a-6a98-11e5-9718-949556da0368.png)

Since the docs http://api.jquery.com/jQuery.extend present the function as a way to mix objects, I would never expect this kind of behaviour. When passing a non-object argument (which can always happen when not having added some sort of check before calling extend), I would expect it to be ignored/skipped; e.g. adding an integer is not reflected in the result of an extend operation:

![2015-10-04 13 05 02](https://cloud.githubusercontent.com/assets/4307137/10267532/98e01a78-6a98-11e5-91ad-249631d8b570.png)

Of course you can sell this as a feature, but I'd rather say this can cause hard to spot bugs…

-Felix

**Comment by mzgol** [comment]
Thanks for the report. However, this is exactly how the ES6 `Object.assign` handles strings; try this in a modern browser:
```js
Object.assign({a: 3}, 'hi'); // Object {0: "h", 1: "i", a: 3}
```

We don't want to deviate from the standard so this needs to stay as it is. If you want it changed, you need to convince TC39, the ECMAScript Standards Committee.

**Comment by felixhayashi** [comment]
Hi @mzgol

> However, this is exactly how the ES6 Object.assign handles strings;

Thanks for the reference.

> If you want it changed, you need to convince TC39, the ECMAScript Standards Committee.

Sure, I'll do that.

-Felix

**Comment by rwaldron** [comment]
We can just settle this here. The [`target` and `...sources` will always be boxed via `ToObject(...)`](http://www.ecma-international.org/ecma-262/6.0/#sec-object.assign) (see: 1 & 5.b.i), this is non-negotiable. If you then review `[ToObject](http://www.ecma-international.org/ecma-262/6.0/#sec-toobject), you'll see the following entry in Table 13: 

| Argument Type | Result |
| --------------------- | ----------- |
| String | Return a new String object whose [[StringData]] internal slot is set to the value of argument. See 21.1 for a description of String objects |


In "code" terms, that's the same as: 

```js
var primitive = "Hello";
var string = new String(primitive);
string;  // String { 0: "H", 1: "e", 2: "l", 3: "l", 4: "o", length: 5 }
```

**Comment by felixhayashi** [comment]
Thanks for the explanation @rwaldron. If you regard this from a plain JS/ECMAScript perspective, it makes a lot of sense and you guys are certainly more aqcuainted with js than I am.

I was just saying that when working with jQuery and reading in the api docs that the `extend` function requires objects as arguments that a string is not naturally thought of as an object in the JS world (`typeof string !== "object"`) so I expected it to be ignored. Because jQuery is often understood as an abstraction/simplification of plain JS, I thought it would be responsible to mention that the function's result has the potential to surprise api users – but I also understand @mzgol's argument to conform to the standard so it's fine…

Regards
Felix

**Comment by dmethvin** [comment]
>  so I expected it to be ignored

Not sure what you mean by this--did you expect invalid API calls to essentially be a no-op? If we haven't documented what happens when you call a jQuery API with some arguments, the behavior by definition is undefined.

**Comment by felixhayashi** [comment]
@dmethvin 

> Not sure what you mean by this--did you expect invalid API calls to essentially be a no-op? If we haven't documented what happens when you call a jQuery API with some arguments, the behavior by definition is undefined.

I know that the outcome is not defined when passing invalid arguments. I simply stated that slicing the string and adding character indeces as properties is probably the least that (unadvanced) api users expect and that it would be imo less problematic to simply ignore the argument in this case. This is my personal opinion, I understood the contra arguments and I did not want to offend anyone so feel free to ignore this thread.
